astropy/utils/misc.py

```diff
@@ -801,3 +802,96 @@ def __init__(cls, cls_name, bases, members):
 
         super(OrderedDescriptorContainer, cls).__init__(cls_name, bases,
                                                         members)
+
+
+class ShapedLikeNDArray(object):
+    """Mixin class to provide shape-changing methods.
+
+    These assume that the class proper defines
+    _replicate(method, *args, **kwargs)
```

Comment:
Yes, I thought about making it an ABC as well. It has the advantage that I could also define `ndim` and `shape` as abstract properties, since these are needed for some of the methods.